From stable-bounces@linux.kernel.org Tue Feb 27 11:14:04 2007
From: Michal Wrobel <xmxwx@asn.pl>
Date: Tue, 27 Feb 2007 11:12:45 -0800 (PST)
Subject: Don't add anycast reference to device multiple times
To: stable@kernel.org
Cc: bunk@stusta.de
Message-ID: <20070227.111245.26276999.davem@davemloft.net>

From: Michal Wrobel <xmxwx@asn.pl>

[IPV6]: anycast refcnt fix

This patch fixes a bug in Linux IPv6 stack which caused anycast address
to be added to a device prior DAD has been completed. This led to
incorrect reference count which resulted in infinite wait for
unregister_netdevice completion on interface removal.

Signed-off-by: Michal Wrobel <xmxwx@asn.pl>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 net/ipv6/addrconf.c |    2 ++
 1 file changed, 2 insertions(+)

--- linux-2.6.20.1.orig/net/ipv6/addrconf.c
+++ linux-2.6.20.1/net/ipv6/addrconf.c
@@ -469,6 +469,8 @@ static void dev_forward_change(struct in
 			ipv6_dev_mc_dec(dev, &addr);
 	}
 	for (ifa=idev->addr_list; ifa; ifa=ifa->if_next) {
+		if (ifa->flags&IFA_F_TENTATIVE)
+			continue;
 		if (idev->cnf.forwarding)
 			addrconf_join_anycast(ifa);
 		else
